nncf/common/quantization/structs.py

```diff
@@ -0,0 +1,175 @@
+        return "B:{bits} M:{mode} SGN:{signedness} PC:{per_channel}".format(
+            bits=self.num_bits,
+            mode='S' if self.mode == QuantizationMode.SYMMETRIC else 'A',
+            signedness='ANY' if self.signedness_to_force is None else ('S' if self.signedness_to_force else 'U'),
+            per_channel='Y' if self.per_channel else 'N')
+
+    def __hash__(self):
+        return hash(str(self))
+
+    def is_valid_requantization_for(self, other: 'QuantizerConfig') -> bool:
+        #pylint:disable=too-many-return-statements
+        """Quantizer config A is a valid requantization for quantizer config B if A is more strict -
+        specifically, it might be reasonable to put quantizer A after quantizer B in tensor data control flow, so that
+        the requantization will further constrain the input tensor data w.r.t. values it can take, but
+        putting quantizer A after quantizer B would be unreasonable."""
+        if self.num_bits > other.num_bits:
+            return False
+        if self.mode is QuantizationMode.ASYMMETRIC and other.mode is \
+                QuantizationMode.SYMMETRIC:
+            return False
+        if self.signedness_to_force is None and other.signedness_to_force is not None:
+            return False
+        if self.signedness_to_force is True and other.signedness_to_force is False:
+            return False
+
+        return True
+
+    def compatible_with_a_unified_scale_linked_qconfig(self, linked_qconfig: 'QuantizerConfig'):
+        return self.num_bits == linked_qconfig.num_bits and \
+               self.mode == linked_qconfig.mode and \
+               self.signedness_to_force == linked_qconfig.signedness_to_force and \
+               self.per_channel == linked_qconfig.per_channel
+
+    def is_a_bitwidth_variant(self, other_qconfig: 'QuantizerConfig') -> bool:
+        return self.per_channel == other_qconfig.per_channel and \
+               self.signedness_to_force == other_qconfig.signedness_to_force and \
+               self.mode == other_qconfig.mode
+
+
+class QuantizerSpec:
+    """A specific (potentially framework-aware) parameter struct required to initialize a
+    given object that performs quantization of an input tensor."""
+    def __init__(self, num_bits: int,
+                 mode: QuantizationMode,
+                 signedness_to_force: bool,
+                 narrow_range: bool):
+        self.num_bits = num_bits
+        self.mode = mode
+        self.signedness_to_force = signedness_to_force
+        self.narrow_range = narrow_range
+
+    def __eq__(self, other: 'QuantizerSpec'):
+        return self.__dict__ == other.__dict__
+
+    @classmethod
+    def from_config(cls, qconfig: QuantizerConfig, narrow_range: bool):
+        return cls(qconfig.num_bits,
+                   qconfig.mode,
+                   qconfig.signedness_to_force,
+                   narrow_range)
+
+
+class QuantizerSetupType(Enum):
+    PATTERN_BASED = "pattern_based"
```

Comment:
Do we plan to support `PATTERN_BASED` approach? Do you really need `QuantizerSetupType` in the common part?

Reply:
Will preserve status quo in this PR and remove the pattern-based mode separately.